tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -0,0 +1,265 @@
+
+        /* Client sets up a resumption connection with the received session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, cb_session_data, cb_session_data_len));
+
+        /* Negotiate second connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_FALSE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+
+        /* Send and receive some more application data. */
+        EXPECT_SUCCESS(s2n_send(server_conn, server_message, sizeof(server_message), &blocked));
+        EXPECT_SUCCESS(s2n_recv(client_conn, client_buffer, sizeof(client_buffer), &blocked));
+        EXPECT_BYTEARRAY_EQUAL(client_buffer, server_message, sizeof(server_message));
+
+        /* Prepare client and server for a third connection */
+        EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+        EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Client sets up a resumption connection with the received session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, cb_session_data, cb_session_data_len));
+
+        /* Negotiate third connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_FALSE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    }
+
+    /* Test: A valid TLS1.3 ticket with TLS1.2 client falls back to a TLS1.2 connection */
+    {
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.3 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+
+        uint8_t server_message[] = "server message";
+        uint8_t client_buffer[sizeof(server_message)];
+        s2n_blocked_status blocked = 0;
+
+        /* Send and receive some application data. This will ensure that the client actually receives the
+        * New Session Ticket post-handshake message. */
+        EXPECT_SUCCESS(s2n_send(server_conn, server_message, sizeof(server_message), &blocked));
+        EXPECT_SUCCESS(s2n_recv(client_conn, client_buffer, sizeof(client_buffer), &blocked));
+        EXPECT_BYTEARRAY_EQUAL(client_buffer, server_message, sizeof(server_message));
```

Comment:
You might want to factor this out into a helper method, since it appears multiple times and is a little unintuitive.